### ISSUE TITLE
group-supergroup transition

### ISSUE THREAD

**Issue by qzhu2017** [qzhu2017/PyXtal#120]
@kevinparrish1 We will start to work on this next week.

- [x] access the supergroup information 
```python
from pyxtal.symmetry import Group 
a=Group(4)
s=a.get_min_supergroup()

>>> s.keys()
dict_keys(['supergroup', 'transformation', 'relations'])
>>> s['supergroup']
[11, 14, 17, 18, 18, 19, 19, 19, 20, 26, 29, 31, 33, 36, 76, 78, 169, 170, 173]
>>> s['relations'][1]
[['2a'], ['2a'], ['2a'], ['2a'], ['2a', '2a']]
```

It should be very similar to group-subgroup transition, except that we need to do two things,

- [x] wyckoff comparability check. Suppose we have a structure from H to G, if the valid WP splitting exists
- [x] For the accepted H-G, we obtain the mapping between the WP sites. Then, we check the difference in order to symmetrize the structure from H to G.
- [x] inverse operation
- [x] the permutation of lattice (e.g., the Pmmm, Fmmm, no way to fix the special lattice)
- [x] support the change of lattice
- [x] support the splitting into two different sets
- [x] add mask to cell translation 
- [x] a script to do subgroup and then test the can find the supergroup backward
- [x] reduce the function calls on Group()
- [x] improve the atomic mapping function
- [x] the splitting from 1 to 3 sites
- [x] clean up the `symmetrize` and `symmetrize_dist` function
- [x] does check_compability return some repetitive solutions?
- [x] Find_mapping a9bd483
- [x] Fix symmetrize dist
- [x] fix the special cases of splitting from 1 to n

**Comment by qzhu2017** [comment]
Checking the compatibility can take a long time.

```python
from pyxtal import pyxtal
s = pyxtal()
s.from_seed('pyxtal/database/cifs/bug.vasp')
print(s)
strucs = s.supergroup(d_tol=1.0, group_type='t')
print(strucs)
```

This will take a long long time to find the possible solutions. Needs to figure out why?

**Comment by qzhu2017** [comment]
@kevinparrish1 

I am working on a new class called supergroups, this class will start from a given pyxtal structure (`H`), and attempt to find the supergroup structure  with a desired symmetry (`G`)

Ideally, I would like to split them into two steps
- `search_paths`, try to find all possible connections between G and H
- `struc_along_path`, for each path, perform the supergroup search. 
-  Whenever it successfully ends up with the desired super group symmetry, we will terminate the calculation.

Now, I have the `struc_along_path` function done. Can you work on `search_paths` function.

**Comment by qzhu2017** [comment]
## find_mapping function
When a possible splitting solution is assigned, there exist multiple solutions to do the atomic mapping. For instance
atoms in G has: {}
atoms in H has: {}
Currently, we used a very straight approach to do the atomic mapping, this sometimes returns many unnecessary solutions. Will be good to reduce the searching cost here.

**Comment by qzhu2017** [comment]
after 6b3a9f5, the supergroup function should be able to handle most cases except:
- too many solutions for a given H-G pair (very rare)
- Too many possible atomic mappings for a given solution (sometimes)
- transition involves a very irregular lattice (e.g., monoclinic with an inclination angle like 130 degree)
- splitting of 1 site to 3 sites (needs to check how many of them exists in reality)

**Comment by qzhu2017** [comment]
@kevinparrish1 I just fixed the issue due to find_mapping. However, I got the following runtime error.
```
Traceback (most recent call last):
  File "pyxtal/miscellaneous/single_check.py", line 34, in <module>
    sup = supergroups(struc_h, G=G, d_tol=0.2, show=True, max_per_G=500)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 491, in __init__
    strucs, w_path, valid = self.struc_along_path(p)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 542, in struc_along_path
    solutions = my.search_supergroup(self.d_tol, self.max_per_G)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 651, in search_supergroup
    mae, disp, mapping, sp = self.get_displacement(G, id, sol, d_tol*1.1)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 751, in get_displacement
    dist, disp, mask = self.symmetrize_dist(splitter, mapping, None, None, d_tol)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 1025, in symmetrize_dist
    coords11_2 += trans12
```

**Comment by qzhu2017** [comment]
Further improve the mapping for the following cases
```
------Crystal from subgroup------
Dimension: 3
Composition: Cu4S14Fe8
Group: Pmm2 (25)
orthorhombic lattice:   6.1300   9.8100  11.4000  90.0000  90.0000  90.0000
Wyckoff sites:
	 S @ [ 0.0000  0.2450  0.0970], WP:  2g, Site symmetry: m..
	 S @ [ 0.5000  0.2547  0.8970], WP:  2h, Site symmetry: m..
	 S @ [ 0.0000  0.5000  0.8954], WP:  1b, Site symmetry: mm2
	 S @ [ 0.5000  0.5000  0.1046], WP:  1d, Site symmetry: mm2
	 S @ [ 0.0000  0.0000  0.8990], WP:  1a, Site symmetry: mm2
	 S @ [ 0.5000  0.0000  0.1007], WP:  1c, Site symmetry: mm2
	 S @ [ 0.0000  0.1614  0.5874], WP:  2g, Site symmetry: m..
	 S @ [ 0.5000  0.1682  0.4160], WP:  2h, Site symmetry: m..
	 S @ [ 0.0000  0.5000  0.4205], WP:  1b, Site symmetry: mm2
	 S @ [ 0.5000  0.5000  0.5832], WP:  1d, Site symmetry: mm2

------Crystal from Seed------
Dimension: 3
Composition: Cu4S14Fe8
Group: Pmma (51)
orthorhombic lattice:   6.1300   9.8100  11.4000  90.0000  90.0000  90.0000
Wyckoff sites:
	 S @ [ 0.2500  0.2500  0.1000], WP:  4k, Site symmetry: m..
	 S @ [ 0.2500  0.5000  0.9000], WP:  2f, Site symmetry: mm2
	 S @ [ 0.2500  0.0000  0.9000], WP:  2e, Site symmetry: mm2
	 S @ [ 0.2500  0.1667  0.5833], WP:  4k, Site symmetry: m..
	 S @ [ 0.2500  0.5000  0.4167], WP:  2f, Site symmetry: mm2
```
Here `2f` and `2f` can be grouped and `4k` can be grouped, so that the total number of permutations can be largely reduced.

**Comment by qzhu2017** [comment]
@kevinparrish1 
How is your progress here? I just several commits and it should work for most of the cases now. However, I am still waiting for your fix on the case of 1 to 3 splitting.

**Comment by kevinparrish1** [comment]
I've been working on the k split for 1->3 all day today. I'm trying to
piece together how everything works through examples.

**Comment by qzhu2017** [comment]
```python
paras = (["R-3", 147, 'k'],
         ["FD3", 70, 't'], #needs 1->3
         ["Fd3m", 166, 't'],
         ["R-3c", 15, 't'],
         ["R32", 5, 't'],
         ["Pm3", 47, 't'], #needs 1->3, still fails
         #["Pmma", 25, 't'], #fails due to atomic mapping
        )
```
@kevinparrish1 I just committed the new test scripts. It is great that you can do the supergroup for `FD3`. However, the case of `Pm3` still fails. Please make sure this example can be fixed soon.

**Comment by kevinparrish1** [comment]
I believe I may have fixed the issue with the merging of 3 positions.

**Comment by qzhu2017** [comment]
@kevinparrish1 Great. I just checked it and it seems that we passed all cases of merging 3 positions. Now, we have the following examples left.

```python
from glob import glob
import numpy as np
import pymatgen.analysis.structure_matcher as sm
from pymatgen import Structure
from pyxtal import pyxtal
from pyxtal.supergroup import supergroups

paras = (
         ["P4_332", 155, 't'], #1-4 splitting
         ["I-43m", 160, 't'], #mapping
         ['I4_132',98,'t'], #mapping
         #["P4_332", 96, 't'],
         #["R-3", 147, 'k'],
         #["P3_112", 5, 't'],
         #["P6_422", 21, 't'],
         #["FD3", 70, 't'], 
         #["Fd3m", 166, 't'],
         #["R-3c", 15, 't'],
         #["R32", 5, 't'],
         #["Pm3", 47, 't'], 
         #["Pmma", 25, 't'],
        )

for para in paras:
    name, H, gtype = para
    name = "pyxtal/miscellaneous/cifs/" + name + ".cif"
    print(name)
    s = pyxtal()
    s.from_seed(name)
    pmg_s1 = s.to_pymatgen()
    G = s.group.number
    for i in range(10):
        struc_h = s.subgroup_once(eps=0.15, H=H, group_type=gtype, mut_lat=False, max_cell=4)
        print(struc_h)
        print(s)
        sup = supergroups(struc_h, G=G, d_tol=0.2, show=True, max_per_G=500)
        if sup.strucs is not None:
            match = False
            for struc in sup.strucs:
                pmg_g = struc.to_pymatgen()
                if sm.StructureMatcher().fit(pmg_g, pmg_s1):
                    match = True
                    break
            if not match:
                print("==Wrong: cannot recover the original structure")
        else:
            print(sup)
            print("==Error: cannot generate structure")
            break
```
Can you work on one of the followings
-  `["P4_332", 155, 't'], #1-4 splitting`?
- `["Pmma", 25, 't'],` #atomic mapping function

**Comment by kevinparrish1** [comment]
I have uploaded a new version of symmetrize_dist that can merge any n positions to 1.

**Comment by qzhu2017** [comment]
> I have uploaded a new version of symmetrize_dist that can merge any n positions to 1.

`["P4_332", 155, 't'], #1-4 splitting` is still not working.

```
Traceback (most recent call last):
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 283, in find_xyz
    G2_basis_xyz=np.linalg.solve(rot_G2,b)
  File "<__array_function__ internals>", line 5, in solve
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/numpy/linalg/linalg.py", line 381, in solve
    _assert_stacked_square(a)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/numpy/linalg/linalg.py", line 204, in _assert_stacked_square
    raise LinAlgError('Last 2 dimensions of the array must be square')
numpy.linalg.LinAlgError: Last 2 dimensions of the array must be square

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pyxtal/miscellaneous/single_check.py", line 43, in <module>
    sup = supergroups(struc_h, G=G, d_tol=0.2, show=True, max_per_G=500)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 569, in __init__
    strucs, w_path, valid = self.struc_along_path(p)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 620, in struc_along_path
    solutions = my.search_supergroup(self.d_tol, self.max_per_G)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 729, in search_supergroup
    mae, disp, mapping, sp = self.get_displacement(G, id, sol, d_tol*1.1)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 829, in get_displacement
    dist, disp, mask = self.symmetrize_dist(splitter, mapping, None, None, d_tol)
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 1114, in symmetrize_dist
    G2_xyz.append(find_xyz(corresponding_ops[j],coord_G2[j],quadrant))
  File "/Users/qiangzhu/anaconda3/lib/python3.8/site-packages/pyxtal-0.2.0-py3.8.egg/pyxtal/supergroup.py", line 293, in find_xyz
    raise RuntimeError('unable to find free parameters using this operation')
RuntimeError: unable to find free parameters using this operation

```

**Comment by kevinparrish1** [comment]
My bad, I think I ran the wrong line. It runs now I believe

**Comment by qzhu2017** [comment]
@kevinparrish1 It works now. But sometimes, it fails and the returns the following msg:
```
Transition to super group: Unsuccessful, check your input
==Error: cannot generate structure
```

**Comment by qzhu2017** [comment]
@kevinparrish1 
You also need to debug the symmetrize function. Right now it returns the correct structure, but wrong distance mapping.
```
Valid structure 214
Au   0.1237 -0.1237  0.0000 ->   0.1250 -0.1250  0.0000 ->   0.0000  0.0000  0.0000   0.0000
Te   0.7663  0.5153  0.6412 ->  -0.2338 -0.4838 -0.3588 ->  -0.0013  0.0022  0.0000   0.0265
Ag   0.1186  0.2500  0.1250 ->   0.1195 -0.7500  0.1250 ->  -0.0004  0.0013  0.0000   0.0138
Ag   0.0000  0.0000  0.2441 ->   0.1195 -0.7500  0.1250 ->   0.1182  0.2513 -0.1191   3.1361
cell:   0.0013 -0.0013  0.0000, disp (A):   3.1361

------Crystal from supergroup  0.061------
Dimension: 3
Composition: Au8Ag24Te16
Group: I4132 (214)
cubic lattice:  10.3800  10.3800  10.3800  90.0000  90.0000  90.0000
Wyckoff sites:
	Au @ [ 0.1250  0.1250  0.1250], WP:  8a, Site symmetry: . 3 2
	Te @ [ 0.7662  0.7662  0.7662], WP: 16e, Site symmetry: . 3 .
	Ag @ [ 0.8801  0.0000  0.2500], WP: 24f, Site symmetry: 2 . .
```

**Comment by qzhu2017** [comment]
@kevinparrish1 Have you fixed the remaining issues on this subject?

**Comment by kevinparrish1** [comment]
I spent all of Friday and most of yesterday working on finishing up the
final problem with the 4 position merging. I realized after a while that
the symmetrize_dist actually works OK. What has actually happened with that
example is that the wyckoff_split returned a splitter object that is
oriented in the wrong direction. Essentially, everything in the splitter is
matched well and works, but some of the translation components of the
orbits have the wrong sign which is causing the errors (ex: a translation
is [1/4, 1/2, 1/2] but should be [-3/4,1/2,1/2]). This is an issue that I
thought I had fixed while working on wyckoff_split, but it seems my
solution had only applied to a few specific scenarios and was not true for
all cases. I will have to review wyckoff_split again to figure out how to
find the correct orientation of the positions. The supergroup.py was
working though.
